businessRules/donations.js

```diff
@@ -1,54 +1,67 @@
 const { v4 } = require('uuid');
 const { createDonation } = require('../models/donations');
-const { getAllIdeas } = require('../models/ideas');
-const { getIdeas } = require('../utils/storage');
+const { findIdea } = require('../validations/ideas');
+const {
+    validateSum,
+    findIdeaWithTotalDonationSum,
+} = require('../validations/donations.js');
+const { findUser } = require('../validations/users');
+const { getUsers } = require('../utils/storage');
 
-function createDonation(firstName, sum, userId, ideaId) {
-    let ideas = getIdeas();
-    const idea = ideas.find((idea) => ideaId === idea.id);
+function createDonationByUnregisteredUser(sum, userId, ideaId) {
+    findIdea(ideaId);
+    validateSum(sum);
 
-    if (!idea) {
-        throw new Error('Error. No idea with such ID found.');
+    let users = getUsers();
+    const user = users.find((user) => userId === user.id);
+
+    let firstName;
+
+    if (!user) {
+        const generateId = v4();
+        const anonymous = 'Anonymous';
+
+        firstName = firstName || anonymous;
+        userId = userId || generateId;
     }
+    createDonation(firstName, sum, userId, ideaId);
+}
 
-    const generateId = v4();
-    const anonymous = 'Anonymous';
+function createDonationByRegisteredUser(sum, userId, ideaId) {
+    findIdea(ideaId);
+    validateSum(sum);
 
-    firstName = firstName || anonymous;
-    userId = userId || generateId;
+    const searchResults = findUser(userId);
```

Comment:
Why `findUser` function returns an array? The name suggests a single user and the parameter that you're passing is a single user id - so I would expect`findUser` function to return a single value, not an array of users?

Reply:
I thought that it would be more conveniet, but while fixing that problem, understood that it complicates everything.Fixed.